File: backends/dotnet/shared/Implementations/Caching/Distributed/DistributedCache.Redis/Handlers/U/Set.cs
// -----------------------------------------------------------------------
// <copyright file="Set.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.DistributedCache.Redis.Handlers.U;

using System.Text.Json;
using D2.Shared.Handler;
using D2.Shared.I18n;
using D2.Shared.Result;
using D2.Shared.Utilities.Serialization;
using Google.Protobuf;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;
using S = D2.Shared.Interfaces.Caching.Distributed.Handlers.U.IUpdate;

/// <summary>
/// Handler for setting a value in the distributed cache.
/// </summary>
///
/// <typeparam name="TValue">
/// The type of the value to cache.
/// </typeparam>
public partial class Set<TValue> : BaseHandler<
        S.ISetHandler<TValue>, S.SetInput<TValue>, S.SetOutput>,
    S.ISetHandler<TValue>
{
    private readonly IConnectionMultiplexer r_redis;

    /// <summary>
    /// Initializes a new instance of the <see cref="Set{TValue}"/> class.
    /// </summary>
    ///
    /// <param name="redis">
    /// The Redis connection multiplexer.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public Set(
        IConnectionMultiplexer redis,
        IHandlerContext context)
        : base(context)
    {
        r_redis = redis;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<S.SetOutput?>> ExecuteAsync(
        S.SetInput<TValue> input,
        CancellationToken ct = default)
    {
        try
        {
            // Serialize the value.
            var bytes = input.Value is IMessage message
                ? message.ToByteArray()
                : JsonSerializer.SerializeToUtf8Bytes(
                    input.Value,
                    SerializerOptions.SR_IgnoreCycles);

            // Connect to Redis and set the value.
            var db = r_redis.GetDatabase();
            await db.StringSetAsync(
                input.Key,
                bytes,
                input.Expiration is { } ttl ? ttl : Expiration.Default);

            // Return success.
            return D2Result<S.SetOutput?>.Ok(
                new S.SetOutput());
        }
        catch (RedisException ex)
        {
            LogSetFailed(Context.Logger, ex, input.Key, TraceId);

            return D2Result<S.SetOutput?>.ServiceUnavailable(traceId: TraceId);
        }
        catch (JsonException ex)
        {
            LogSetSerializationFailed(Context.Logger, ex, input.Key, TraceId);

            return D2Result<S.SetOutput?>.UnhandledException(
                messages: [TK.Common.Errors.COULD_NOT_BE_SERIALIZED],
                traceId: TraceId);
        }

        // Let the base handler catch any other exceptions.
    }

    /// <summary>
    /// Logs that a Redis exception occurred while setting a cached value.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "RedisException occurred while setting value for key '{Key}'. TraceId: {TraceId}")]
    private static partial void LogSetFailed(ILogger logger, Exception ex, string key, string? traceId);

    /// <summary>
    /// Logs that a JSON serialization exception occurred while setting a cached value.
    /// </summary>
    [LoggerMessage(EventId = 2, Level = LogLevel.Error, Message = "JsonException occurred while serializing value for key '{Key}'. TraceId: {TraceId}")]
    private static partial void LogSetSerializationFailed(ILogger logger, Exception ex, string key, string? traceId);
}

File: backends/node/shared/implementations/caching/distributed/redis/src/handlers/u/set.ts
import type Redis from "ioredis";
import { BaseHandler, type IHandlerContext, type RedactionSpec } from "@d2/handler";
import { TK } from "@d2/i18n";
import { D2Result, ErrorCodes, HttpStatusCode } from "@d2/result";
import { redisErrorResult } from "../../redis-error-result.js";
import type { DistributedCache } from "@d2/interfaces";
import { JsonCacheSerializer, type ICacheSerializer } from "../../serialization.js";

export class Set<TValue>
  extends BaseHandler<DistributedCache.SetInput<TValue>, DistributedCache.SetOutput>
  implements DistributedCache.ISetHandler<TValue>
{
  private readonly redis: Redis;
  private readonly serializer: ICacheSerializer<TValue>;

  constructor(redis: Redis, context: IHandlerContext, serializer?: ICacheSerializer<TValue>) {
    super(context);
    this.redis = redis;
    this.serializer = serializer ?? new JsonCacheSerializer<TValue>();
  }

  override get redaction(): RedactionSpec {
    return { inputFields: ["value"] };
  }

  protected async executeAsync(
    input: DistributedCache.SetInput<TValue>,
  ): Promise<D2Result<DistributedCache.SetOutput | undefined>> {
    try {
      let serialized: string | Buffer;
      try {
        serialized = this.serializer.serialize(input.value);
      } catch (err: unknown) {
        this.context.logger.warn(
          `Cache serialization failed: ${err instanceof Error ? err.message : String(err)}`,
        );
        return D2Result.fail({
          messages: [TK.common.errors.COULD_NOT_BE_SERIALIZED],
          statusCode: HttpStatusCode.InternalServerError,
          errorCode: ErrorCodes.COULD_NOT_BE_SERIALIZED,
        });
      }

      if (input.expirationMs !== undefined) {
        await this.redis.set(input.key, serialized, "PX", input.expirationMs);
      } else {
        await this.redis.set(input.key, serialized);
      }

      return D2Result.ok({ data: {} });
    } catch (err: unknown) {
      return redisErrorResult(err);
    }
  }
}
